### DISCUSSION TITLE
Setup - PostgresML

### DISCUSSION THREAD

**Discussion by giscus** [postgresml/postgresml#541] [Announcements]
# Setup - PostgresML

Train and deploy models to make online predictions using only SQL, with an open source Postgres extension.

https://postgresml.org/user_guides/transformers/setup/

**Comment by dventimihasura** [comment]
Two things.

First, running the installation commands as shown with `sudo` causes this warning:

```shell
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
```

Is it necessary to use `sudo`?  Installation instructions that cause the system to emit warnings will worry people kicking the tires and will scare some of them away.

Second, and more important, after following the instructions to the letter, I get this error when trying to use transformers.

```sql
psql (14.6 (Ubuntu 14.6-0ubuntu0.22.04.1))
Type "help" for help.

postgres=# SELECT pgml.transform(
    'translation_en_to_fr',
    inputs => ARRAY[
        'Welcome to the future!',
        'Where have you been all this time?'
    ]
) AS french;
ERROR:  called `Result::unwrap()` on an `Err` value: PyErr { type: <class 'ModuleNotFoundError'>, value: ModuleNotFoundError("No module named 'transformers'"), traceback: Some(<traceback object at 0x7fbaea7dc700>) }
CONTEXT:  src/bindings/transformers.rs:19:62
```

This is for postgresql-14 and pgml-14 on Ubuntu 22.

**Comment by montanalow** [comment]
1) Unfortunately, python dependency management is fraught with peril and installing in the system Python is required for Postgres to be able to find the modules for most generally distributed versions.

![image](https://user-images.githubusercontent.com/276125/218291703-8f729f80-4332-4bc4-b0c8-3daf2bba880c.png)

There may be ways we could get around this by using individual package managers (e.g. apt on debian/ubuntu) to install Python dependencies system wide, but we'd need documentation for every distribution and OS. I think a better path forward would be to sandbox the Postgres plPython dependencies (or at least just our own) using a virtualenv for [py03](https://github.com/PyO3/pyo3). 

2) I think you need to also install the transformers pip package (and probably torch and/or tensorflow). I'll see what docs need to be updated.

**Comment by ahmedrachid** [comment]
Hello, would you please provide guidance on how to install postgresML extension on CentOS ? Many thanks !

**Comment by montanalow** [comment]
Hi @ahmedrachid, we don't maintain a CentOS package, although you can convert the `deb` package we distribute with [alien](https://wiki.debian.org/Alien). Otherwise, you'll need to build from source.

**Reply by ahmedrachid** [reply]
Oh i see.. I will give it a try then, many thanks for your answer. I'll be testing it in a distributed MPP PostgreSQL, do you think that model training can be distributed ?

**Reply by montanalow** [reply]
PostgresML training happens in a single connection process, so it will not work out of the box to distribute training throughout a Postgres-XL cluster. Parallelization happens within the training algorithm, although training can be offloaded to a replica.

**Reply by ahmedrachid** [reply]
That's great to know, many thanks for your help :)

**Reply by ahmedrachid** [reply]
Hello @montanalow , can you please share with me the "deb" package link ? I'm unable to find it and would like to use Alien to convert it to RPM. Many thanks
